woocommerce/woocommerce-gateway-stripe
# Fix Stripe Link saved payment methods

Fixes #3224

On top of fixing that this PR also includes some improvements to the general Stripe Link saved payment method flow.

## Changes proposed in this Pull Request:

This PR fixes 4 things with Stripe Link saved payment methods: 

1. When the customer cache is cleared, we weren't clearing the saved Link payment methods. This would cause Link payment methods to not actually be saved until the cache would naturally expire (24 hours).
2. Fix the display of Link Payment methods.
3. Subscriptions purchased with a Stripe Link payment method weren't displaying the correctly on the My Account Subscription details. 
4. Unable to add saved payment method. You would get the following error. 

**MY ACCOUNT PAYMENT METHODS ERROR**
<p align="center">
<img width="600" alt="Screenshot 2024-07-24 at 12 42 50 PM" src="https://github.com/user-attachments/assets/7a34f401-6cd3-48ca-9e9f-7933fe6caf8b">
</p> 

---

**LINK TOKEN DISPLAY**
| Checkout | My Account |
|--------|--------|
| <img width="690" alt="Screenshot 2024-07-24 at 3 09 21 PM" src="https://github.com/user-attachments/assets/5655051f-f736-4017-928e-29eacd3d8ada"><p align="center"><sup><strong>Before:</strong> Stripe Link tokens displayed default value</sup></p> | <img width="858" alt="Screenshot 2024-07-24 at 3 04 53 PM" src="https://github.com/user-attachments/assets/b78134b5-74f0-49a4-b51c-7647df8b366f"><p align="center"><sup><strong>Before:</strong> Stripe Link tokens missing value</sup></p> |
| <img width="698" alt="Screenshot 2024-07-24 at 3 11 36 PM" src="https://github.com/user-attachments/assets/c90ce8c1-b62a-4bf1-b854-bd0b864650d5"><p align="center"><sup><strong>After:</strong> Stripe Link tokens now displayed their display name (like shortcode checkout)</sup></p> | <img width="843" alt="Screenshot 2024-07-24 at 3 03 32 PM" src="https://github.com/user-attachments/assets/ffcc7424-dcf7-4000-a592-93003b60c694"><p align="center"><sup><strong>After:</strong> Stripe Link tokens now display the brand</sup></p>| 

---
**SUBSCRIPTION LINK PM DISPLAY**

| Before | After |
|--------|--------|
| <img width="844" alt="Screenshot 2024-07-24 at 3 56 20 PM" src="https://github.com/user-attachments/assets/6d20d9bb-d699-4b12-a4c6-02042d7fdc46"> | <img width="831" alt="Screenshot 2024-07-24 at 3 55 21 PM" src="https://github.com/user-attachments/assets/1646ba36-716e-4198-b9b7-bbfbfaa06968"> | 

## Testing instructions

> [!note]
> You will need to be logged into Link and have a stored payment method to test this PR.

1. Enable Stripe Link in your Stripe payment method settings
3. Go to the **My Account > Payment methods** page. 
4. Add a payment method.
5. Attempt to add a card.
   - On develop you will get the mandate required error.
   - On this branch there shouldn't be any error.
6. View the My Account > Payment methods page. 
7. Note that the Stripe Link token you just added has the **Stripe Link** name. 
8. Add a product to your cart.
9. Go to the block checkout and confirm that the token also displays there with the email address used to create the Stripe Link payment method. 

---


**Post merge**
